cEP-0022.md

```diff
@@ -0,0 +1,162 @@
+
+From what is evident, the brute-force is going to take a huge amount of time
+to run especially on large projects. For this the `coala-quickstart` will
+have a `lite-mode`. We assume that the lowest level in which there can be
+variations will be file types as a starting point, we try to guess the
+settings not from the entire project, but for each file type, we grab a set
+of files at random and run coala and again look for *green* settings, if we
+find a value for a setting as *"green"*, we add it to the config file, if
+not we drop the setting/bear. This mode is error prone, but is a lot
+faster than the brute force way. **NOTE:** The successive runs of the
+`lite-mode` should build upon the data from the previous runs to generate
+less error prone `.coafile` (less error prone over here indicates not likely
+to generate errors in the project), so that running `lite-mode` a number of
+times is still faster than Brute Force. We choose this method as it is highly
+unlikely that different files among the same file types have a different
+set of codestyle rules in a project.
+There can be endless number of possibilities and assumptions:
+
+- Files having similar kind of names can be having different codestyles.
+- Files in a given directory can be following different codestyles.
+- Different parts of files maybe even following a different codestyle depending
+  on function name which can also be handled by placing `ignore`s at
+  appropriate places in the file dict, such kind of functionality can only be
+  added to the bears themselves, but instead we stop our train of thought over
+  here.
+
+## Learning from real world projects
+
+We save ourselves the additional work for building some feature that may
+hardly ever be used by any org, instead we stop assumming and start
+implementing and learning from our available resources.
+
+We now choose a list of orgs for which we try to test our `green-mode`
+ourselves. We check whether brute force is taking too much time or whether
+the `lite-mode` is producing the correct results (i.e. *green* results). We
+choose these orgs in such a way that they are very famous or very well known or
+act as upstream repositories for a huge amount of orgs. This way a very large
+number of orgs may be mimicking the code style of these orgs itself and if
+our further optimizations can fix the problem of generating *green* config
+files for these organizations, we are in turn solving the problem for these
+other orgs/communities at the same time.
+
+We learn manually from these orgs, what combinations of these bear-setting
+values are they using, whether they are using it all over their project or only
+in certain scenarious, whether they have different codestyles in different
+directories or different nomenclature of file patterns require a different
+set of settings in their project. We give `coala-quickstart`, the ability
+to learn and identify these patterns, recognizing these scenarios for further
```

Comment:
Such kind of "learning" would require storing data permanently somewhere, right?

Reply:
the sentence is ill-framed 😅, learning is gonna be done by us as I don't presume that coala will become a hot-shot in a very short amount of time, most of these tasks will be done by members of our own org, running quickstart green mode, checking inconsistencies, making a PR in other orgs, which will require us to adapt `coala-quickstart` for further identification of such norms whenever reported to us.

Reply:
So we currently have a "recommended" set of bears. Maybe there could different packages of recommendations? Like, Google has its own code guidelines, the PSF has its own different set of guidelines, blah blah. Maybe orgs could choose to align themselves with one of these "packages".

I vaguely remember discussing this some time ago with @sils, not sure if it was ever implemented.

Reply:
definitely should be working towards having definitions of pre-existing pre-defined coding style guidelines.  We havent started that, and there are issues about this, and this project probably cant get anywhere near achieving that, as the matrix between style guidelines and bears is too complicated - The `aspects` area is intended to be that matrix, and we have an `aspects` project this year which should bring us closer to that goal, but we wouldnt want to tie this project to the aspects project - too risky .